File: services/worker-service/src/budget_scopes.py
"""Categories AI spend is budgeted and accounted against.

Mirror of ``apps/api/src/common/constants/budget-scopes.ts``. The worker
cannot import TypeScript, so this list is duplicated and CI-locked by
``tests/test_budget_scopes.py``.

Every LLM call the worker makes must carry one of these as its ``scope``:
it selects which per-category cap rag-service enforces, which Redis usage
counters are incremented, and which row the spend lands on in
``budget_ledger``.
"""

from __future__ import annotations

from typing import Final

BUDGET_SCOPES: Final[tuple[str, ...]] = (
    "case_digest",
    "doctrine_extract",
    "mcq_question",
    "essay_prompt",
    "flashcard",
    "subject_outline",
    "subject_classification",
    "bar_exam_answer",
    "ai_answer",
)

# Named constants so a task references a symbol rather than retyping the
# string — the ad-hoc literals in worker payloads are exactly what let
# four of these categories drift out of the ledger entirely.
SCOPE_CASE_DIGEST: Final[str] = "case_digest"
SCOPE_DOCTRINE_EXTRACT: Final[str] = "doctrine_extract"
SCOPE_MCQ_QUESTION: Final[str] = "mcq_question"
SCOPE_ESSAY_PROMPT: Final[str] = "essay_prompt"
SCOPE_FLASHCARD: Final[str] = "flashcard"
SCOPE_SUBJECT_OUTLINE: Final[str] = "subject_outline"
SCOPE_SUBJECT_CLASSIFICATION: Final[str] = "subject_classification"
SCOPE_BAR_EXAM_ANSWER: Final[str] = "bar_exam_answer"
SCOPE_AI_ANSWER: Final[str] = "ai_answer"


def is_budget_scope(value: object) -> bool:
    """True when ``value`` is one of the known budget scopes."""
    return isinstance(value, str) and value in BUDGET_SCOPES

File: apps/web/src/features/admin/budget-scopes.ts
/**
 * Mirror of `apps/api/src/common/constants/budget-scopes.ts`.
 *
 * The web app cannot import from the API package, so the list is
 * duplicated and CI-locked by `budget-scopes.test.ts`, which reads the
 * TypeScript source directly. Keep the two in the same order.
 */
export const BUDGET_SCOPES = [
  'case_digest',
  'doctrine_extract',
  'mcq_question',
  'essay_prompt',
  'flashcard',
  'subject_outline',
  'subject_classification',
  'bar_exam_answer',
  'ai_answer',
] as const;

export type BudgetScope = (typeof BUDGET_SCOPES)[number];

/** Operator-facing names for the admin budget table. */
export const BUDGET_SCOPE_LABELS: Record<BudgetScope, string> = {
  case_digest: 'Case digests',
  doctrine_extract: 'Doctrine extracts',
  mcq_question: 'MCQ questions',
  essay_prompt: 'Essay prompts',
  flashcard: 'Flashcards',
  subject_outline: 'Subject outlines',
  subject_classification: 'Subject classification',
  bar_exam_answer: 'Bar exam answers',
  ai_answer: 'AI answers',
};

export function budgetScopeLabel(scope: string): string {
  return (BUDGET_SCOPE_LABELS as Record<string, string>)[scope] ?? scope;
}
